cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DocumentStore.java

```diff
@@ -0,0 +1,192 @@
+     * key will open or create an encrypted database. If opening a database, the key used to
+     * create the database must be used. If encryption is not enabled for this platform, returning
+     * a non-null key will result in an exception.
+     * </p>
+     * @param location The location on the file system where the underlying files should be stored.
+     *                 Must be a directory.
+     * @param provider KeyProvider object. Use a {@link NullKeyProvider} if the database shouldn't
+     *                 be encrypted.
+     * @return An existing or newly created store.
+     * @throws DatastoreNotCreatedException if the database cannot be opened
+     */
+    public static DocumentStore getInstance(File location, KeyProvider provider) throws DatastoreNotCreatedException {
+        try {
+            synchronized (documentStores) {
+                DocumentStore ds = documentStores.get(location);
+                boolean created = false;
+                if (ds == null) {
+                    created = checkPathAndCreateIfNeeded(location);
+                    ds = new DocumentStore(location, provider);
+                    documentStores.put(location, ds);
+                    eventBus.post(new DatabaseOpened(ds.databaseName));
+                }
+                if (created) {
+                    eventBus.post(new DatabaseCreated(ds.databaseName));
+                }
+                return ds;
+            }
+        }
+        catch (IOException e) {
+            throw new DatastoreNotCreatedException("Database not found: " + location, e);
+        } catch (SQLException e) {
+            throw new DatastoreNotCreatedException("Database not initialized correctly: " + location, e);
+        } catch (DatastoreException e) {
+            throw new DatastoreNotCreatedException("Datastore not initialized correctly: " + location, e);
+        }
+    }
+
+    private static boolean checkPathAndCreateIfNeeded(File location) {
+        logger.fine("Datastore path: " + location);
+
+        boolean created = false;
+        if(!location.exists()){
+            created = location.mkdir();
+        }
+
+        Misc.checkArgument(location.isDirectory(), "Input path is not a valid directory");
+        Misc.checkArgument(location.canWrite(), "Datastore directory is not writable");
+        return created;
+    }
+
+    @SuppressWarnings("unchecked")
+    public void close() {
+        synchronized (documentStores) {
+            DocumentStore ds = documentStores.remove(location);
+            if (ds != null) {
+                ((DatabaseImpl)database).close();
+                ((IndexManagerImpl)query).close();
+            }
+            else {
+                throw new IllegalStateException("DocumentStore "+location+" already closed");
+            }
+        }
+        eventBus.post(new DatabaseClosed(databaseName));
+    }
+
+    public void delete() throws IOException {
```

Comment:
It seems a bit inconsistent that `delete()` will throw an `IOException` for some problem (e.g. file locked), but the failure to create or open for the same reason would throw a specific `Datastore*` exception.